File: example/android/app/src/main/kotlin/com/exponea/example/managers/CustomerTokenStorage.kt
package com.exponea.example.managers

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.exponea.example.MainActivity
import com.exponea.sdk.util.Logger
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.google.gson.reflect.TypeToken
import java.util.concurrent.TimeUnit
import kotlin.math.abs

/**
 * !!! WARN for developers.
 * This implementation is just proof of concept. Do not rely on any part of it as possible solution for Customer Token
 * handling.
 * It is in your own interest to provide proper token generating and handling of its cache, expiration and secured storing.
 */
class CustomerTokenStorage(
    private var networkManager: NetworkManager = NetworkManager(),
    private val gson: Gson = Gson(),
    private val prefs: SharedPreferences = MainActivity.APP_CONTEXT!!.getSharedPreferences("FlutterSharedPreferences", Context.MODE_PRIVATE)
) {

    companion object {
        private const val CUSTOMER_TOKEN_CONF = "CustomerTokenConf"
        val INSTANCE = CustomerTokenStorage()
    }

    private var currentConf: Config? = null

    private var tokenCache: String? = null

    private var lastTokenRequestTime: Long = 0

    fun retrieveJwtToken(): String? {
        validateChangedConf()
        val now = System.currentTimeMillis()
        if (TimeUnit.MILLISECONDS.toMinutes(abs(now - lastTokenRequestTime)) < 5) {
            // allows request for token once per 5 minutes, doesn't care if cache is NULL
            Logger.d(this, "[CTS] Token retrieved within 5min, using cache $tokenCache")
            return tokenCache
        }
        lastTokenRequestTime = now
        if (tokenCache != null) {
            // return cached value
            Logger.d(this, "[CTS] Token cache returned $tokenCache")
            return tokenCache
        }
        synchronized(this) {
            // recheck nullity just in case
            if (tokenCache == null) {
                tokenCache = loadJwtToken()
            }
        }
        return tokenCache
    }

    private fun validateChangedConf() {
        val storedConf = loadConfiguration()
        if (currentConf != storedConf) {
            // reset token
            tokenCache = null
            lastTokenRequestTime = 0
        }
        currentConf = storedConf
    }

    private fun loadJwtToken(): String? {
        currentConf = loadConfiguration()
        val host = currentConf?.host
        val projectToken = currentConf?.projectToken
        val publicKey = currentConf?.publicKey
        val customerIds = currentConf?.customerIds
        if (
            host == null || projectToken == null ||
            publicKey == null || customerIds == null || customerIds.size == 0
        ) {
            Logger.d(this, "[CTS] Not configured yet")
            return null
        }
        val reqBody = hashMapOf(
            "project_id" to projectToken,
            "kid" to publicKey,
            "sub" to customerIds
        )
        val jsonRequest = gson.toJson(reqBody)
        val response = networkManager.post(
            "$host/webxp/exampleapp/customertokens",
            null,
            jsonRequest
        ).execute()
        Logger.d(this, "[CTS] Requested for token with $jsonRequest")
        if (!response.isSuccessful) {
            if (response.code == 404) {
                // that is fine, only some BE has this endpoint
                Logger.d(this, "[CTS] Token request returns 404")
                return null
            }
            Logger.e(this, "[CTS] Token request returns ${response.code}")
            return null
        }
        val jsonResponse = response.body?.string()
        val responseData = try {
            gson.fromJson(jsonResponse, Response::class.java)
        } catch (e: Exception) {
            Logger.e(this, "[CTS] Token cannot be parsed from $jsonResponse")
            return null
        }
        if (responseData?.token == null) {
            Logger.e(this, "[CTS] Token received NULL")
        }
        Logger.d(this, "[CTS] Token received ${responseData?.token}")
        return responseData?.token
    }

    private fun loadConfiguration(): Config {
        return Config(
            // See config.dart how are stored fields
            host = prefs.getString("flutter.base_url", null),
            projectToken = prefs.getString("flutter.project_token", null),
            publicKey = prefs.getString("flutter.advanced_auth_token", null),
            // See home.dart how are stored customerIds in _identifyCustomer method
            customerIds = prefs.getString("flutter.customer_ids", null).let {
                gson.fromJson(it ?: "{}", object : TypeToken<HashMap<String, String>>() {}.type)
            }
        )
    }

    data class Response(
        @SerializedName("customer_token")
        val token: String?,

        @SerializedName("expire_time")
        val expiration: Int?
    )

    data class Config(
        var host: String? = null,
        var projectToken: String? = null,
        var publicKey: String? = null,
        var customerIds: HashMap<String, String>? = null

    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Config

            if (host != other.host) return false
            if (projectToken != other.projectToken) return false
            if (publicKey != other.publicKey) return false
            if (customerIds != other.customerIds) return false

            return true
        }

        override fun hashCode(): Int {
            var result = host?.hashCode() ?: 0
            result = 31 * result + (projectToken?.hashCode() ?: 0)
            result = 31 * result + (publicKey?.hashCode() ?: 0)
            result = 31 * result + (customerIds?.hashCode() ?: 0)
            return result
        }
    }
}

File: example/ios/Runner/CustomerTokenStorage.swift
//
//  CustomerTokenStorage.swift
//  Runner
//
//  Created by Adam Mihalik on 02/03/2023.
//

import Foundation
import ExponeaSDK

class CustomerTokenStorage {

    public static var shared = CustomerTokenStorage()

    private let semaphore: DispatchQueue = DispatchQueue(
        label: "CustomerTokenStorageLockingQueue",
        attributes: .concurrent
    )
    
    private var currentConf: Config?

    private var tokenCache: String?

    private var lastTokenRequestTime: Double = 0

    func retrieveJwtToken() -> String? {
        validateChangedConf()
        let now = Date().timeIntervalSince1970
        let timeDiffMinutes = abs(now - lastTokenRequestTime) / 60.0
        if timeDiffMinutes < 5 {
            // allows request for token once per 5 minutes, doesn't care if cache is NULL
            return tokenCache
        }
        lastTokenRequestTime = now
        if tokenCache != nil {
            // return cached value
            return tokenCache
        }
        semaphore.sync(flags: .barrier) {
            // recheck nullity just in case
            if tokenCache == nil {
                tokenCache = loadJwtToken()
            }
        }
        return tokenCache
    }
    
    private func validateChangedConf() {
        let storedConf = loadConfiguration()
        if currentConf != storedConf {
            // reset token
            tokenCache = nil
            lastTokenRequestTime = 0
        }
        currentConf = storedConf
    }

    private func loadJwtToken() -> String? {
        currentConf = loadConfiguration()
        guard let host = currentConf?.host,
              let projectToken = currentConf?.projectToken,
              let publicKey = currentConf?.publicKey,
              let customerIds = currentConf?.customerIds,
              customerIds.count > 0 else {
            Exponea.logger.log(.verbose, message: "CustomerTokenStorage not configured yet")
            return nil
        }
        guard let url = URL(string: "\(host)/webxp/exampleapp/customertokens") else {
            Exponea.logger.log(.error, message: "Invalid URL host \(host) for CustomerTokenStorage")
            return nil
        }
        var requestBody: [String: Codable] = [
            "project_id": projectToken,
            "kid": publicKey,
            "sub": customerIds
        ]
        do {
            let postData = try JSONSerialization.data(withJSONObject: requestBody, options: [])
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.httpBody = postData
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
            let httpSemaphore = DispatchSemaphore(value: 0)
            var httpResponseData: Data?
            var httpResponse: URLResponse?
            let httpTask = URLSession.shared.dataTask(with: request) {
                httpResponseData = $0
                httpResponse = $1
                _ = $2
                httpSemaphore.signal()
            }
            httpTask.resume()
            _ = httpSemaphore.wait(timeout: .distantFuture)
            if let httpResponse = httpResponse as? HTTPURLResponse {
                // this is optional check - we looking for 404 posibility
                switch httpResponse.statusCode {
                case 404:
                    // that is fine, only some BE has this endpoint
                    return nil
                case 300..<599:
                    Exponea.logger.log(
                        .error,
                        message: "Example token receiver returns \(httpResponse.statusCode)"
                    )
                    return nil
                default:
                    break
                }
            }
            guard let httpResponseData = httpResponseData else {
                Exponea.logger.log(.error, message: "Example token response is empty")
                return nil
            }
            let responseData = try JSONDecoder().decode(TokenResponse.self, from: httpResponseData)
            if responseData.token == nil {
                Exponea.logger.log(.error, message: "Example token received NULL")
            }
            return responseData.token
        } catch let error {
            Exponea.logger.log(.error, message: "Example token cannot be parsed due error \(error.localizedDescription)")
            return nil
        }
    }
    
    private func loadConfiguration() -> Config {
        let prefs = UserDefaults.standard
        var customerIds: [String: String]?
        if let customerIdsString = prefs.string(forKey: "flutter.customer_ids"),
           let customerIdsData = customerIdsString.data(using: .utf8),
           let loadedIds = try? JSONDecoder().decode([String: String].self, from: customerIdsData) {
            customerIds = loadedIds
        }
        return Config(
            // See config.dart how are stored fields
            host: prefs.string(forKey: "flutter.base_url"),
            projectToken: prefs.string(forKey: "flutter.project_token"),
            publicKey: prefs.string(forKey: "flutter.advanced_auth_token"),
            // See home.dart how are stored customerIds in _identifyCustomer method
            customerIds: customerIds
        )
    }
}

struct TokenResponse: Codable {
    var token: String?
    var expiration: Int?

    enum CodingKeys: String, CodingKey {
        case token = "customer_token"
        case expiration = "expire_time"
    }
}

struct Config: Equatable {
    var host: String?
    var projectToken: String?
    var publicKey: String?
    var customerIds: [String: String]?
}
